core/modules/replica/Controller.cc

```diff
@@ -0,0 +1,837 @@
+/*
+ * LSST Data Management System
+ * Copyright 2017 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+// Class header
+#include "replica/Controller.h"
+
+// System headers
+#include <iostream>
+#include <stdexcept>
+#include <sys/types.h>  // pid_t
+#include <unistd.h>     // getpid()
+
+// Qserv headers
+#include "lsst/log/Log.h"
+#include "replica/Common.h"
+#include "replica/DatabaseServices.h"
+#include "replica/DeleteRequest.h"
+#include "replica/FindRequest.h"
+#include "replica/FindAllRequest.h"
+#include "replica/Messenger.h"
+#include "replica/Performance.h"
+#include "replica/ReplicationRequest.h"
+#include "replica/ServiceManagementRequest.h"
+#include "replica/ServiceProvider.h"
+#include "replica/StatusRequest.h"
+#include "replica/StopRequest.h"
+
+// This macro to appear witin each block which requires thread safety
+#define LOCK_GUARD std::lock_guard<std::mutex> lock(_mtx)
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.Controller");
+
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+//////////////////////////////////////////////////////////////////////////
+//////////////////////////  RequestWrapperImpl  //////////////////////////
+//////////////////////////////////////////////////////////////////////////
+
+/**
+ * Request-type specific wrappers
+ */
+template <class  T>
+struct RequestWrapperImpl
+    :   ControllerRequestWrapper {
+
+    /// The implementation of the vurtual method defined in the base class
+    virtual void notify() {
+        if (_onFinish == nullptr) { return; }
+        _onFinish(_request);
+    }
+
+    RequestWrapperImpl(typename T::pointer const& request,
+                       typename T::callback_type onFinish)
+        :   ControllerRequestWrapper(),
+            _request(request),
+            _onFinish(onFinish) {
+    }
+
+    /// Destructor
+    ~RequestWrapperImpl() override = default;
+
+    /// Implement a virtual method of the base class
+    std::shared_ptr<Request> request() const override {
+        return _request;
+    }
+
+private:
+
+    // The context of the operation
```

Comment:
these vars contain the context of the operation? I don't understand the comment as written.

Reply:
This was an artifact. Removed.